[AArch64] Initial compiler support for SVE unwind on Windows.

efriedma-quic opened the pull request:
Most bits of this are straightforward: when we emit SVE instructions in the prologue/epilogue, emit corresponding opcodes.

The unfortunately nasty bit is the handling of the frame pointer in functions that use the SVE calling convention.  If we have SVE callee saves, and need to restore the stack pointer from the frame pointer, it's impossible to encode callee saves that happen after the frame pointer. So this patch rearranges the stack to put SVE callee saves first.  This isn't really that complicated on its own, but it leads to a lot of tricky conditionals (see FPAfterSVECalleeSaves).

efriedma-quic committed:
[AArch64] Initial compiler support for SVE unwind on Windows.

Most bits of this are straightforward: when we emit SVE instructions in
the prologue/epilogue, emit corresponding opcodes.

The unfortunately nasty bit is the handling of the frame pointer in
functions that use the SVE calling convention.  If we have SVE
callee saves, and need to restore the stack pointer from the frame
pointer, it's impossible to encode callee saves that happen after the
frame pointer. So this patch rearranges the stack to put SVE callee
saves first.  This isn't really that complicated on its own, but it
leads to a lot of tricky conditionals (see FPAfterSVECalleeSaves).

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2757,10 +2853,27 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
-    ScalableOffset = -SVEStackSize;
-  if (!UseFP && (isFixed || isCSR))
-    ScalableOffset = SVEStackSize;
+  bool FPAfterSVECalleeSaves =
```
`getFrameIndexReferenceFromSP` also needs to be updated to handle this layout (currently `-pass-remarks-analysis=stack-frame-layout` reports the normal layout).

efriedma-quic replied:
Fixed getFrameIndexReferenceFromSP; still need to add a test.  Haven't looked at resolveFrameOffsetReference for SVE CSs yet.

efriedma-quic replied:
Added a test for stack-frame-layout.  Added the assert for resolveFrameOffsetReference (as you note, a frame index should never refer to a CSR in this context).

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2757,10 +2853,27 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
-    ScalableOffset = -SVEStackSize;
-  if (!UseFP && (isFixed || isCSR))
-    ScalableOffset = SVEStackSize;
+  bool FPAfterSVECalleeSaves =
```
With `"frame-pointer"="all"`, I think the addresses of SVE locals are incorrect. For example, the `f4` test function resolves the offset as:
```
	sub	x1, x29, #8
	mov	x0, sp
	addvl	x1, x1, #-18
```
Which is subtracting the size of the SVE callee-saves that are above the FP.

efriedma-quic replied:
Fixed.

mstorsjo reviewed (approved):
Looking good overall, thanks! A couple minor inline comments.

mstorsjo commented on llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:
```diff
@@ -6057,24 +6067,24 @@ void llvm::emitFrameOffset(MachineBasicBlock &MBB,
-  assert(!(NeedsWinCFI && (NumPredicateVectors || NumDataVectors)) &&
-         "WinCFI not supported with SVE vectors");
+  assert(!(NeedsWinCFI && NumPredicateVectors) &&
+         "WinCFI can't allocate fractions of an SVE data vector");
```
Also, it's not clear what/where is being done, so that when we have one or more predicate vectors to back up, we allocate space for them in the form of full data vectors. For the general calling convention, all of p4-p11 are backed up I presume, but what if we'd be clobbering only one of them in a function where we don't need to back them all up? How is that handled on e.g. Linux today?

MacDue replied:
The SVE stack size is always rounded up to a 16-byte multiple (implicitly multiple by vscale), so we always allocate some number of full SVE vectors (which also ensures the stack stays 16-byte aligned). I don't think predicate-sized allocations can occur in frame lowering.

mstorsjo replied:
Right... What I'm asking is, is it possible to hit this assert? Looking at `decomposeStackOffsetForFrameOffsets` it does seem like it would be possible to end up with a nonzero `NumPredicateVectors` here?

MacDue replied:
In `decomposeStackOffsetForFrameOffsets()` if `NumPredicateVectors % 8 == 0` (which will be the case if the allocation has been aligned to 16-bytes), it'll change the allocation to be `NumDataVectors =  NumPredicateVectors / 8` (and no predicate vectors). So I don't think this can be hit from frame lowering. I think it may be used for some other stack allocations (but I think `NeedsWinCFI` is false for those).

mstorsjo replied:
Right, I see - thanks, that answers my question!

MacDue reviewed (commented):
A few initial comments:

efriedma-quic committed:
Partially address review comments.

efriedma-quic commented on llvm/test/CodeGen/AArch64/win-sve.ll:
```diff
@@ -0,0 +1,1034 @@
+; CHECK-NEXT:    .seh_allocz 17
+; CHECK-NEXT:    .seh_endepilogue
+; CHECK-NEXT:    ret
+                i64 %n5, i64 %n6, i64 %n7, i64 %n8, i64 %n9) personality ptr @__CxxFrameHandler3 {
```
Fixed.

efriedma-quic commented on llvm/test/CodeGen/AArch64/win-sve.ll:
```diff
@@ -0,0 +1,1034 @@
+; CHECK-NEXT:    .seh_endproc
+  call void asm "", "~{d8}"()
+  ret void
+}
```
Added.  (This space is allocated by getFixedObjectSize().  f6 actually contains the same kind of allocation, although it's maybe not obvious at first glance.)

efriedma-quic committed:
More review comment fixes.

efriedma-quic committed:
Fix formatting

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2757,10 +2871,25 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
+      else
+        ScalableOffset = SVECalleeSavedStack;
+    } else if (!UseFP && (isFixed || isCSR)) {
+      ScalableOffset = SVEStackSize;
```
Again, this looks like it could be incorrect for fixed CSRs, which are `SVEStackSize - SVECalleeSavedStack` above the SP.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2757,10 +2871,25 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
+    if (UseFP) {
+      if (!(isFixed || isCSR))
+        ScalableOffset = SVECalleeSavedStack - SVEStackSize;
+      else
+        ScalableOffset = SVECalleeSavedStack;
```
Can you think of any way to write a test involving callee-saves in resolveFrameOffsetReference?  I think maybe debug info calls into this code, but I don't think the result is relevant for the actual debug info.  And I don't think anything else refers to callee-saves like this: the prologue/epilogue code doesn't resolve offsets like this, and nothing else has any reason to refer to them.

I guess the fix is to change the `isFixed || isCSR` to just `isFixed`?

MacDue replied:
I added the assert ` assert(!(AFI->getSVECalleeSavedStackSize() && isCSR));` and it seems this can get hit by `AArch64RegisterInfo::eliminateFrameIndex`, which seems to need the correct offset. It only fails for one test though (`llvm/test/CodeGen/AArch64/stack-hazard.ll`, `@svecc_csr_d8_allocd`).

efriedma-quic replied:
Pushed a fix.

The frame index in question is not actually a CSR, but there's open space in the CSR range due to alignment, so it gets allocated as part of the "CSR" area.  And the calculation was in fact broken in this case.

efriedma-quic committed:
Fix handling of "CSR" references.

efriedma-quic committed:
Actually fix handling of "CSR" references.

Previous version didn't non-CSR SVE allocations correctly.

efriedma-quic commented:
@MacDue ping

MacDue commented:
> @MacDue ping

I'll take another careful look look tomorrow :+1:

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2116,9 +2180,11 @@ void AArch64FrameLowering::emitPrologue(MachineFunction &MF,
-    assert(IsSVECalleeSave(CalleeSavesBegin) && "Unexpected instruction");
-    while (IsSVECalleeSave(MBBI) && MBBI != MBB.getFirstTerminator())
-      ++MBBI;
+    if (!FPAfterSVECalleeSaves) {
```
I think this and the following `if (!FPAfterSVECalleeSaves) ` could be folded into one region like:
https://gist.github.com/MacDue/eff5ffa8cc735c1b56e5987555bf3506 (which also avoids having the incorrect `CalleeSavesBegin` variable when `FPAfterSVECalleeSaves` is active).

MacDue reviewed (commented):
I think this mostly looks okay (I can't spot anything broken :slightly_smiling_face: -- but this is pretty tricky code). It is missing some of the testing we have for the non-Windows code path, e.g., https://github.com/llvm/llvm-project/blob/main/llvm/test/CodeGen/AArch64/framelayout-sve.mir, which includes tests with fixed-stack objects (which I don't think are tested by `win-sve.ll`). It may be worth making some equivalent tests for Windows.

efriedma-quic committed:
Address review comments.

Fixed issue lower tailcc and related constructs.

Code style changes.

efriedma-quic commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2303,10 +2371,16 @@ void AArch64FrameLowering::emitEpilogue(MachineFunction &MF,
   bool CombineAfterCSRBump = false;
-  if (!CombineSPBump && PrologueSaveSize != 0) {
+  if (FPAfterSVECalleeSaves) {
+    AfterCSRPopSize = FixedObject;
```
Fixed this to +=, and added a test.

Also addressed other review comments.

efriedma-quic replied:
Adapted those tests for Windows.  Let me know if that covers what you're concerned about, or if we need something else.

Discovered a couple of issues with predicate registers in the process...

efriedma-quic committed:
More fixes.

Add more tests.  Correctly spill p12-p15.  Fix issue spilling an odd
number of predicates.

MacDue reviewed (approved):
LGTM, thanks for addressing my comments :+1:. I'm not entirely certain that everything is exactly right here (I don't personally work on Windows targets), but I don't think this should have any impact on SVE for non-Windows targets, so in that sense, I consider this a fairly safe change. 

Please wait a few days to see if anyone else has comments before landing.

efriedma-quic committed:
Add hazard padding error.

mstorsjo commented:
With this merged, I run into issues with projects that detect support for SVE, which previously bailed out.

The main issue is hitting issues with codeview debug info; I saw that you have a hack for this at https://github.com/llvm/llvm-project/commit/5e3fa842d15a0be130dcd2b6671a0176f30af8c3. You'd need to extend this to cover the predicate registers too; I'm hitting this with `P0` in one repro.

If compiling without optimization, I'm hitting another assert relating to codeview debug info too - I presume this is the other unresolved detail you mentioned in the commit message in that hack.

Repro:
```c
typedef __attribute__((neon_vector_type(16))) signed char int8x16_t;
typedef __SVInt8_t svint8_t;
typedef __SVBool_t svbool_t;
__attribute__((__clang_arm_builtin_alias(__builtin_sve_svdup_n_s8))) svint8_t
svdup_n_s8(char);
void svsel_s8(svbool_t, svint8_t, svint8_t);
__attribute__((__clang_arm_builtin_alias(__builtin_sve_svundef_s8))) svint8_t
svundef_s8();
svbool_t svwhilelt_b8(int, int);
__attribute__((__clang_arm_builtin_alias(__builtin_sve_svset_neonq_s8)))
svint8_t svset_neonq_s8(svint8_t, int8x16_t);
int8x16_t saoCuStatsE1_sve2_endY_in;
void saoCuStatsE1_sve2_endY() {
  svbool_t svpred = svwhilelt_b8(0, 0);
  svsel_s8(svpred, svset_neonq_s8(svundef_s8(), saoCuStatsE1_sve2_endY_in),
           svdup_n_s8(3));
}
```
```console
$ clang -target aarch64-windows-msvc -c repro.c -march=armv9-a -g -O2
fatal error: error in backend: unknown codeview register P0
$ clang -target aarch64-windows-msvc -c repro.c -march=armv9-a -g
clang: ../lib/CodeGen/AsmPrinter/CodeViewDebug.cpp:1281: void llvm::CodeViewDebug::collectVariableInfoFromMFTable(llvm::DenseSet<std::pair<const llvm::DINode*, const llvm::DILocation*> >&): Assertion `!FrameOffset.getScalable() && "Frame offsets with a scalable component are not supported"' failed.
```

efriedma-quic commented:
Those two crashes are the ones I've seen, yes, and are addressed by the hack.

The issue is that the projects detect SVE without debug info, then try to build with debug info?

So... I have the following options here right now:

- Add a temporary disabled-by-default flag for SVE on Windows, so projects can't detect the SVE support.
- Add some form of debug info hack: either intentionally add incorrect mappings, or somehow drop the bits of debug info that would depend on those mappings.

Longer-term, we're reaching out to Microsoft for more information here, but not sure how long that will take.

mstorsjo commented:
> Those two crashes are the ones I've seen, yes, and are addressed by the hack.

Oh, right, the other issue is also silenced by the hack - sorry, I had missed that.

> The issue is that the projects detect SVE without debug info, then try to build with debug info?

In this case, I think they do apply the same debug info options both when detecting and building, the problem is that the issue with debug info doesn't show up on a trivial sized detection snippet. I did contribute a big enough snippet that did trigger when we don't support generating unwind info, https://aomedia.googlesource.com/aom/+/5ccdc66ab6eb8eb300eda854fab4ff250b2c2f92%5E%21/, but it doesn't trigger the issues with debug info.

> So... I have the following options here right now:
> 
> * Add a temporary disabled-by-default flag for SVE on Windows, so projects can't detect the SVE support.
> * Add some form of debug info hack: either intentionally add incorrect mappings, or somehow drop the bits of debug info that would depend on those mappings.
> 
> Longer-term, we're reaching out to Microsoft for more information here, but not sure how long that will take.

I can work around it in my test setups temporarily by not enabling codeview debug info for aarch64 windows, but by the time of the 21.x branch (which should be in 5-7 weeks roughly?) I think it would be good to apply either of those kinds of fixes - no strong opinion on which one.

davemgreen commented:
Hi @efriedma-quic, we are still seeing a crash with debug info (on non-windows platforms), as in https://godbolt.org/z/6Eced39b8. It only triggers in debug builds because of some extra calls to resolveFrameOffsetReference that do not occur without debug info. Any ideas? Thanks

MacDue commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -2736,13 +2849,21 @@ StackOffset AArch64FrameLowering::resolveFrameOffsetReference(
+      isTargetWindows(MF) && AFI->getSVECalleeSavedStackSize();
+
   if (isSVE) {
+    assert(-ObjectOffset > (int64_t)AFI->getSVECalleeSavedStackSize() &&
+           "Math isn't correct for CSRs with FPAfterSVECalleeSaves");
```
This assertion should be moved under the `FPAfterSVECalleeSaves` check below to prevent the crash on non-Windows targets. The maths may also need fixing for Windows (that might also hit this), but that's a separate thing.

vtjnash commented on llvm/lib/Target/AArch64/AArch64FrameLowering.cpp:
```diff
@@ -1874,12 +1899,51 @@ void AArch64FrameLowering::emitPrologue(MachineFunction &MF,
+      Subtarget.isTargetWindows() && AFI->getSVECalleeSavedStackSize();
+
+  if (FPAfterSVECalleeSaves && AFI->hasStackHazardSlotIndex())
+    reportFatalUsageError("SME hazard padding is not supported on Windows");
```
Is this fatal error supposed to be usually unreachable? It seems rather easy to accidentally trigger without having any idea what "SME hazard padding" is supposed to mean to the user, or why that means it is a user error to use any `vscale` operations in a function (reduced from clang which made the alloca):
```
$ cat reduced.ll
target datalayout = "e-m:w-p270:32:32-p271:32:32-p272:64:64-p:64:64-i32:32-i64:64-i128:128-n32:64-S128-Fn32"
target triple = "aarch64-w64-windows-gnu"

define void @ssyrk_direct_sme1_2VLx2VL() #0 {
  %1 = alloca <vscale x 16 x i1>, align 2
  store <vscale x 16 x i1> zeroinitializer, ptr %1, align 2
  ret void
}

attributes #0 = { "aarch64_pstate_sm_body" }

$ llc reduced.ll --mtriple=aarch64-w64-mingw32 --mattr=+sve2,+sme
LLVM ERROR: SME hazard padding is not supported on Windows
```

Refs: https://github.com/OpenMathLib/OpenBLAS/issues/5585#issue-3758608878
